source/common/http/filter/extauth.cc

```diff
@@ -0,0 +1,308 @@
+  // who wants to make auth decisions based on the host to which the client
+  // started out trying to talk to.
+  //
+  // We may need to make this configurable later, so I'm leaving this line
+  // in for reference.
+  // reqmsg->headers().insertHost().value(config_->cluster_);
+
+  // After setting up whatever headers we need to, make sure the body is
+  // correctly marked as empty.
+
+  reqmsg->headers().insertContentLength().value(uint64_t(0));
+
+  // Finally, we mark the request as being an Ambassador auth request.
+
+  reqmsg->headers().addReference(header_to_add, value_to_add.get());
+
+  // Fire the request up. When it's finished, we'll get a call to
+  // either onSuccess() or onFailure().
+
+  ENVOY_STREAM_LOG(trace, "ExtAuth contacting auth server", *callbacks_);
+
+  auth_request_ =
+      config_->cm_.httpAsyncClientForCluster(config_->cluster_)
+          .send(std::move(reqmsg), *this, Optional<std::chrono::milliseconds>(config_->timeout_));
+
+  // It'll take some time for our auth call to complete. Stop
+  // filtering while we wait for it.
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+/*
+ * decodeHeaders is called at the point that the HTTP machinery handling
+ * the request has parsed the HTTP body for this request. We don't need
+ * to do anything special here; we just need to make sure that we don't
+ * let things proceed until our auth call is done.
+ */
+
+FilterDataStatus ExtAuth::decodeData(Buffer::Instance&, bool) {
+  if (auth_complete_) {
+    return FilterDataStatus::Continue;
+  }
+  return FilterDataStatus::StopIterationAndBuffer;
+}
+
+/*
+ * decodeTrailers is called at the point that the HTTP machinery handling
+ * the request has parsed the HTTP trailers for this request. We don't need
+ * to do anything special here; we just need to make sure that we don't
+ * let things proceed until our auth call is done.
+ */
+
+FilterTrailersStatus ExtAuth::decodeTrailers(HeaderMap&) {
+  if (auth_complete_) {
+    return FilterTrailersStatus::Continue;
+  }
+  return FilterTrailersStatus::StopIteration;
+}
+
+ExtAuthStats ExtAuth::generateStats(const std::string& prefix, Stats::Scope& scope) {
+  std::string final_prefix = prefix + "extauth.";
+  return {ALL_EXTAUTH_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
+}
+
+/*
+ * onSuccess is called when our asynch auth request succeeds, meaning
+ * "the HTTP protocol was successfully followed to completion" -- it
+ * could still be the case that the auth server gave us a failure
+ * response.
+ */
+
+void ExtAuth::onSuccess(Http::MessagePtr&& response) {
+  // We're done with our auth request, so make sure it gets shredded.
+  auth_request_ = nullptr;
+
+  dumpHeaders("onSuccess", request_headers_);
+
+  // What did we get back from the auth server?
+
+  uint64_t response_code = Http::Utility::getResponseStatus(response->headers());
+  std::string response_body(response->bodyAsString());
```

Comment:
move body? generally fix extra copies of response body/headers in this method as you shouldn't need any

Reply:
👍